py3status/modules/gazelle.py

```diff
@@ -0,0 +1,171 @@
+
+Color options:
+    color_ratio_ok: color when ration is bigger than 1 (default color_good)
+    color_ratio_warning: color when ration is between required ratio and 1
+        (available only when tracker has a dynamic required ratio (default color_warning)
+    color_ratio_ko: color when ratio below required ratio (default color_bad)
+
+@author raspbeguy <raspbeguy@hashtagueule.fr>
+@license MIT
+"""
+
+from py3status.exceptions import RequestException
+
+try:
+    # Python 3
+    from http.cookiejar import (CookieJar, MozillaCookieJar)
+except ImportError:
+    # Python 2
+    from cookielib import (CookieJar, MozillaCookieJar)
+
+
+class Py3status:
+    """
+    """
+
+    # available configuration parameters
+    baseurl = 'https://mytracker.stuff'
+    cache_timeout = 60
+    cookiepath = '/var/tmp/gazelle_ratio_cookie.txt'
+    format = 'Ratio: {ratio}'
+    limit_required = None
+    limit_warning = 1
+    password = ''
+    username = ''
+
+    def _readable_size(self, size):
+        num = size
+        for unit in ['B', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB', 'EiB', 'ZiB']:
+            if abs(num) < 1024.0:
+                return "%3.1f %s" % (num, unit)
+            num /= 1024.0
+        return "%.1f%s" % (num, 'YiB')
+
+    def _login(self):
+        self.py3.request(
+                self.baseurl+"/login.php",
+                data={'username': self.username, 'password': self.password},
+                headers={'User-Agent': 'Mozilla/5.0'},
+                cookiejar=self._cookiejar
+                )
+        try:
+            self._index()  # just to try if login info are good
+        except RequestException:
+            self.py3.log(
+                    "Cannot login to %s" % self.baseurl,
+                    level=self.py3.LOG_ERROR
+                    )
+            raise
+        try:
+            self._cookiejar.save(ignore_discard=True)
+        except IOError:
+            self.py3.log(
+                    "Cannot write cookie at %s" % self.cookiepath,
+                    level=self.py3.LOG_WARNING
+                    )
+        except AttributeError:
+            pass  # That means that no cookie pass has been set
+        else:
+            self.py3.log(
+                    "Created cookie at %s" % self.cookiepath,
+                    level=self.py3.LOG_INFO
+                    )
+
+    def _index(self):
+        resp = self.py3.request(
+                self.baseurl+"/ajax.php",
+                params={'action': "index"},
+                headers={
+                    'User-Agent': 'Mozilla/5.0'
+                    },
+                cookiejar=self._cookiejar
+                )
+        return resp.json()
+
+    def post_config_hook(self):
+        if self.cookiepath:
+            self._cookiejar = MozillaCookieJar(self.cookiepath)
+            try:
+                self._cookiejar.load(ignore_discard=True)
+                self._index()  # just to try if saved cookie is good
```

Comment:
We should skip this and do it in `gazette` instead because this could lead to slow startup for all modules. You can try and test this offline... and maybe just do simple `first_run` loop (see `coin_market` again). Not sure about others.

Reply:
@tobes I kept mentioning this for Internet modules. I take it it is not doable to fix that inside py3status? To have all modules return `''` on first run regardless?

Reply:
The problem is that currently we start each module in the main thread.  We need to move to a situation where each module is started in a thread therefore slow modules do not slow the whole process.  I think fixing that will be easier than enforcing clean/quick starting modules.  It may also make the logic of modules simpler if they can do slow startups.

We need to be careful here as container modules need to have their 'children' started before themselves, but that is not too hard to do (it may also be something that can be removed as a dependency but currently it is an issue)

We now have the structures in place to do this so I may look into it.  One problem is that really we need to make some structural changes and these can be painful to review and merge, but it is something we need to do to make py3status more slim and agile.

I created #1197 for this